src/main/java/org/broadinstitute/hellbender/tools/examples/ExampleStreamingPythonExecutor.java

```diff
@@ -0,0 +1,146 @@
+import org.broadinstitute.hellbender.engine.ReadWalker;
+import org.broadinstitute.hellbender.engine.ReferenceContext;
+import org.broadinstitute.hellbender.exceptions.GATKException;
+import org.broadinstitute.hellbender.utils.python.StreamingPythonScriptExecutor;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+import org.broadinstitute.hellbender.utils.runtime.AsynchronousStreamWriterService;
+
+import java.io.*;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Example ReadWalker program that uses a Python streaming executor to stream summary data from a BAM
+ * input file to a Python process through a FIFO. The read data is accumulated in a List until a batch
+ * size threshold is reached, at which point the batch is handed off to an asynchronous write service,
+ * which writes the batch to the FIFO stream on a background thread. The Python process in turn just
+ * writes the data to an output file.
+ *
+ * <ol>
+ * <li>Opens a FIFO for writing.</li>
+ * <li>Creates an AsyncWriterervice</li>
+ * <li>Writes a string of attributes for each read to the List until the batchSize theshold is reavched.</li>
+ * <li>Uses Python to read each attribute line from the FIFO, and write it to the output file.</li>
+ * </ol>
+ */
+@CommandLineProgramProperties(
+        summary = "Example/toy program that uses a Python script.",
+        oneLineSummary = "Example/toy program that uses a Python script.",
+        programGroup = ExampleProgramGroup.class,
+        omitFromCommandLine = true
+)
+public class ExampleStreamingPythonExecutor extends ReadWalker {
+    private final static String NL = String.format("%n");
+
+    @Argument(fullName = StandardArgumentDefinitions.OUTPUT_LONG_NAME,
+            shortName = StandardArgumentDefinitions.OUTPUT_SHORT_NAME,
+            doc = "Output file")
+    private File outputFile; // output file produced by Python code
+
+    @Argument(fullName = "batchSize",
+            doc = "Size of a batch for writing")
+    private int batchSize = 1000;
+
+    // Create the Python executor. This doesn't actually start the Python process, but verifies that
+    // the requestedPython executable exists and can be located.
+    final StreamingPythonScriptExecutor pythonExecutor = new StreamingPythonScriptExecutor(true);
+
+    private FileOutputStream fifoWriter;
+    private AsynchronousStreamWriterService<String> asyncWriter = null;
+    private List<String> batchList = new ArrayList<>(batchSize);
+    private int batchCount = 0;
+
+    @Override
+    public void onTraversalStart() {
+
+        // Start the Python process, and get a FIFO from the executor to use to send data to Python. The lifetime
+        // of the FIFO is managed by the executor; the FIFO will be destroyed when the executor is terminated.
+        pythonExecutor.start(Collections.emptyList());
+        final File fifoFile = pythonExecutor.getFIFOForWrite();
```

Comment:
Yeah, I have mixed feelings. I agree that its subtle and somewhat non-intuitive. I had a higher-level API in a previous incarnation, but making it general purpose got complicated, and it wound up obscuring the orchestration, which the caller needs to understand because the python code has be structured to conform to it. I'd lobby to wait until we build a couple of real tools before introducing something higher level.